drivers/i2c/i2c_omap.c

```diff
@@ -0,0 +1,765 @@
+		if (k_uptime_get_32() > timeout) {
+			LOG_ERR("Bus busy timeout");
+#ifdef CONFIG_I2C_OMAP_BUS_RECOVERY
+			return i2c_omap_recover_bus(dev);
+#else
+			return -ETIMEDOUT;
+#endif /* CONFIG_I2C_OMAP_BUS_RECOVERY */
+		}
+		k_busy_wait(100);
+	}
+	return 0;
+}
+
+/**
+ * @brief Performs data transfer for the OMAP I2C driver.
+ *
+ * This function is responsible for handling the data transfer logic for the OMAP I2C driver.
+ * It reads the status register and performs the necessary actions based on the status flags.
+ * It handles both receive and transmit logic, and also handles error conditions such as NACK,
+ * arbitration lost, receive overrun, and transmit underflow.
+ *
+ * @param dev Pointer to the device structure.
+ *
+ * @return Returns 0 on success, or a negative error code on failure.
+ */
+static int i2c_omap_transfer_message_ll(const struct device *dev)
+{
+	struct i2c_omap_data *data = DEV_DATA(dev);
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+	uint16_t bits, stat;
+	int err = 0;
+
+	do {
+		bits = i2c_base_addr->IE;
+		stat = i2c_base_addr->STAT;
+		stat |= bits;
+		if (data->receiver) {
+			stat &= ~(I2C_OMAP_STAT_XDR | I2C_OMAP_STAT_XRDY);
+		} else {
+			stat &= ~(I2C_OMAP_STAT_RDR | I2C_OMAP_STAT_RRDY);
+		}
+		if (stat & I2C_OMAP_STAT_NACK) {
+			err |= I2C_OMAP_STAT_NACK;
+			i2c_base_addr->STAT = I2C_OMAP_STAT_NACK;
+		}
+		if (stat & I2C_OMAP_STAT_AL) {
+			err |= I2C_OMAP_STAT_AL;
+			i2c_base_addr->STAT = I2C_OMAP_STAT_AL;
+		}
+		if (stat & I2C_OMAP_STAT_ARDY) {
+			i2c_base_addr->STAT = I2C_OMAP_STAT_ARDY;
+		}
+		if (stat & (I2C_OMAP_STAT_ARDY | I2C_OMAP_STAT_NACK | I2C_OMAP_STAT_AL)) {
+			i2c_base_addr->STAT = I2C_OMAP_STAT_RRDY | I2C_OMAP_STAT_RDR |
+					      I2C_OMAP_STAT_XRDY | I2C_OMAP_STAT_XDR |
+					      I2C_OMAP_STAT_ARDY;
+			break;
+		}
+		/* Handle receive logic */
+		if (stat & (I2C_OMAP_STAT_RRDY | I2C_OMAP_STAT_RDR)) {
+			int buffer = (stat & I2C_OMAP_STAT_RRDY) ? i2c_base_addr->BUF
+								 : i2c_base_addr->BUFSTAT;
+			i2c_omap_transmit_receive_data(dev, buffer);
+			i2c_base_addr->STAT = (stat & I2C_OMAP_STAT_RRDY) ? I2C_OMAP_STAT_RRDY
+									  : I2C_OMAP_STAT_RDR;
+			continue;
+		}
+		/* Handle transmit logic */
+		if (stat & (I2C_OMAP_STAT_XRDY | I2C_OMAP_STAT_XDR)) {
+			int buffer = (stat & I2C_OMAP_STAT_XRDY) ? i2c_base_addr->BUF
+								 : i2c_base_addr->BUFSTAT;
+			i2c_omap_transmit_receive_data(dev, buffer);
+			i2c_base_addr->STAT = (stat & I2C_OMAP_STAT_XRDY) ? I2C_OMAP_STAT_XRDY
+									  : I2C_OMAP_STAT_XDR;
+			continue;
+		}
+		if (stat & I2C_OMAP_STAT_ROVR) {
+			LOG_ERR("Receive overrun");
+			err |= I2C_OMAP_STAT_ROVR;
+			i2c_base_addr->STAT = I2C_OMAP_STAT_ROVR;
+			break;
+		}
+		if (stat & I2C_OMAP_STAT_XUDF) {
+			LOG_ERR("Transmit underflow");
+			err |= I2C_OMAP_STAT_XUDF;
+			i2c_base_addr->STAT = I2C_OMAP_STAT_XUDF;
+			break;
+		}
+		if (!stat) {
+			data->cmd_err |= -EAGAIN;
+			return -EAGAIN;
+		}
+		data->cmd_err = err;
```

Comment:
Having both `cmd_err` and `err` is confusing, sometimes we return one, other times we only add error bits to one, we even assign between them here. How about removing `err`. Everywhere we set it then return, could you just return it directly? Or set it in `data->cmd_err`. The only return code we ever check is for `-EAGAIN`, so why `return err;` below?

Also this function is already called in a loop that waits until stat is set, why nested loops at all? Could you combine the loops by simply removing the do-while loop from this function?

Reply:
Ack
I will change it.

Reply:
Removed the err variable and nested loop

Reply:
You removed the wrong loop, we want to keep the outer loop as it is the one that waits for events when polling. The inner loop simply tried to process all waiting events, but it doesn't itself wait once all events are cleared. Now we only wait once before trying to process all the events. If events are not queued up or fast enough now the transaction will fail. Idea was to still wait on status, but each call to `i2c_omap_transfer_message_ll()` would handle one event. That way the error handling could be simplified, it would just return any error directly.

Reply:
My bad. I totally missed the part.

Reply:
Hello @glneo .If the loop is removed from i2c_omap_transfer_message_ll, it would break the functionality needed for read-after-write, burst read, and burst write operations. The outer loop was essential because it waits for events during polling, allowing these transactions to complete successfully. The inner loop, by contrast, helps in processing multiple R/W I2C operations within a single transactions.

Reply:
Currently the outer-loop polls until at least one bit is set in `STAT`, then the inner-loop will loop until all bits set in `STAT` have been processed and cleared. After that the inner-loop exits and the outer-loop goes back to waiting for more `STAT` bits to come in. There is no need for the inner-loop as `i2c_omap_transfer_message_ll` could simply process one `STAT` bit per call, the outer-loop would then see some other `STAT` bits are still set and go right back to call `i2c_omap_transfer_message_ll` again, nothing is lost by removing the inner-loop.

What we gain by removing it is easier to understand flow control and error handling. Plus better timeout handling. Simply loop until the whole transfer is complete or we go too long without a `STAT` bit being set (timeout). This also makes the future IRQ support easier too, simply call the inner-function for each IRQ (which happens any time a `STAT` bit gets set).

Reply:
I have removed the loop from `i2c_omap_transfer_message_ll`

Reply:
I will be closing this.